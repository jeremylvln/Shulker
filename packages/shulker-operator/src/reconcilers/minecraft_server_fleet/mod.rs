use std::{collections::BTreeMap, sync::Arc, time::Duration};

use futures::StreamExt;
use google_agones_crds::v1::game_server::GameServer;
use k8s_openapi::api::core::v1::ConfigMap;
use kube::{
    api::{ListParams, PatchParams},
    runtime::{
        controller::Action,
        finalizer::{finalizer, Event as Finalizer},
        watcher::Config,
        Controller,
    },
    Api, Client, ResourceExt,
};
use tracing::*;

use shulker_crds::{
    condition::{ConditionStatus, HasConditions},
    v1alpha1::minecraft_server_fleet::{MinecraftServerFleet, MinecraftServerFleetStatus},
};

use self::{
    config_map::ConfigMapBuilder, fleet::FleetBuilder, fleet_autoscaler::FleetAutoscalerBuilder,
};

use super::{builder::reconcile_builder, status::patch_status, ReconcilerError, Result};

mod config_map;
mod fleet;
mod fleet_autoscaler;

static FINALIZER: &str = "minecraftserverfleets.shulkermc.io";

struct MinecraftServerFleetReconciler {
    client: kube::Client,

    // Builders
    config_map_builder: ConfigMapBuilder,
    fleet_builder: FleetBuilder,
    fleet_autoscaler_builder: FleetAutoscalerBuilder,
}

impl MinecraftServerFleetReconciler {
    async fn reconcile(
        &self,
        api: Api<MinecraftServerFleet>,
        minecraft_server_fleet: Arc<MinecraftServerFleet>,
    ) -> Result<Action> {
        reconcile_builder(&self.config_map_builder, minecraft_server_fleet.as_ref()).await?;
        let fleet = reconcile_builder(&self.fleet_builder, minecraft_server_fleet.as_ref()).await?;
        reconcile_builder(
            &self.fleet_autoscaler_builder,
            minecraft_server_fleet.as_ref(),
        )
        .await?;

        if let Some(fleet) = &fleet {
            if let Some(fleet_status) = &fleet.status {
                let mut minecraft_server_fleet = minecraft_server_fleet.as_ref().clone();
                if minecraft_server_fleet.status.is_none() {
                    minecraft_server_fleet.status = Some(MinecraftServerFleetStatus::default());
                }

                let status = minecraft_server_fleet.status.as_mut().unwrap();

                status.replicas = fleet_status.replicas;
                status.ready_replicas = fleet_status.ready_replicas;
                status.allocated_replicas = fleet_status.allocated_replicas;

                if status.ready_replicas > 0 || status.allocated_replicas > 0 {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::True,
                        "AtLeastOneReadyOrAllocated".to_string(),
                        "One or more servers are ready or allocated".to_string(),
                    );
                } else {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::False,
                        "NoneReady".to_string(),
                        "No server are ready".to_string(),
                    );
                };

                patch_status(
                    &api,
                    &PatchParams::apply("shulker-operator"),
                    &minecraft_server_fleet,
                )
                .await?;
            }
        }

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, minecraft_server_fleet: Arc<MinecraftServerFleet>) -> Result<Action> {
        info!(
            name = minecraft_server_fleet.name_any(),
            namespace = minecraft_server_fleet.namespace(),
            "cleaning up MinecraftServerFleet",
        );

        Ok(Action::await_change())
    }

    fn get_common_labels(
        minecraft_server_fleet: &MinecraftServerFleet,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            (
                "app.kubernetes.io/name".to_string(),
                minecraft_server_fleet.name_any(),
            ),
            (
                "app.kubernetes.io/component".to_string(),
                "minecraftserver".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                minecraft_server_fleet.spec.cluster_ref.name.clone(),
            ),
            (
                "minecraftserverfleet.shulkermc.io/name".to_string(),
                minecraft_server_fleet.name_any(),
            ),
        ])
    }
}

#[instrument(skip(ctx, minecraft_server_fleet))]
async fn reconcile(
    minecraft_server_fleet: Arc<MinecraftServerFleet>,
    ctx: Arc<MinecraftServerFleetReconciler>,
) -> Result<Action> {
    let ns = minecraft_server_fleet.namespace().unwrap();
    let minecraft_server_fleets_api: Api<MinecraftServerFleet> =
        Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = minecraft_server_fleet.name_any(),
        namespace = ns,
        "reconciling MinecraftServerFleet",
    );

    finalizer(
        &minecraft_server_fleets_api,
        FINALIZER,
        minecraft_server_fleet,
        |event| async {
            match event {
                Finalizer::Apply(minecraft_server_fleet) => {
                    ctx.reconcile(
                        minecraft_server_fleets_api.clone(),
                        minecraft_server_fleet.clone(),
                    )
                    .await
                }
                Finalizer::Cleanup(minecraft_server_fleet) => {
                    ctx.cleanup(minecraft_server_fleet.clone()).await
                }
            }
        },
    )
    .await
    .map_err(|e| ReconcilerError::FinalizerError(Box::new(e)))
}

fn error_policy(
    _minecraft_server_fleet: Arc<MinecraftServerFleet>,
    error: &ReconcilerError,
    _ctx: Arc<MinecraftServerFleetReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&minecraft_server_fleet, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let minecraft_server_fleets_api = Api::<MinecraftServerFleet>::all(client.clone());
    if let Err(e) = minecraft_server_fleets_api
        .list(&ListParams::default().limit(1))
        .await
    {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = MinecraftServerFleetReconciler {
        client: client.clone(),
        config_map_builder: ConfigMapBuilder::new(client.clone()),
        fleet_builder: FleetBuilder::new(client.clone()),
        fleet_autoscaler_builder: FleetAutoscalerBuilder::new(client.clone()),
    };

    Controller::new(
        minecraft_server_fleets_api,
        Config::default().any_semantic(),
    )
    .owns(
        Api::<ConfigMap>::all(client.clone()),
        Config::default().any_semantic(),
    )
    .owns(
        Api::<GameServer>::all(client.clone()),
        Config::default().any_semantic(),
    )
    .shutdown_on_signal()
    .run(reconcile, error_policy, context.into())
    .filter_map(|x| async move { std::result::Result::ok(x) })
    .for_each(|_| futures::future::ready(()))
    .await;
}
