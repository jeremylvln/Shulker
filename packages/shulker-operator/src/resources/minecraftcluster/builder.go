/*
Copyright (c) Jérémy Levilain
SPDX-License-Identifier: GPL-3.0-or-later
*/

package resources

import (
	"fmt"

	shulkermciov1alpha1 "github.com/jeremylvln/shulker/packages/shulker-crds/src/v1alpha1"
	common "github.com/jeremylvln/shulker/packages/shulker-resource-utils/src"
	"k8s.io/apimachinery/pkg/runtime"
)

type MinecraftClusterResourceBuilder struct {
	Instance *shulkermciov1alpha1.MinecraftCluster
	Scheme   *runtime.Scheme
}

func (b *MinecraftClusterResourceBuilder) ResourceBuilders() ([]common.ResourceBuilder, []common.ResourceBuilder) {
	builders := []common.ResourceBuilder{
		b.MinecraftClusterForwardingSecret(),
		b.MinecraftClusterProxyServiceAccount(),
		b.MinecraftClusterProxyRole(),
		b.MinecraftClusterProxyRoleBinding(),
		b.MinecraftClusterMinecraftServerServiceAccount(),
		b.MinecraftClusterMinecraftServerRole(),
		b.MinecraftClusterMinecraftServerRoleBinding(),
	}
	dirtyBuilders := []common.ResourceBuilder{}

	return builders, dirtyBuilders
}

func (b *MinecraftClusterResourceBuilder) getForwardingSecretName() string {
	return fmt.Sprintf("%s-forwarding-secret", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getProxyServiceAccountName() string {
	return fmt.Sprintf("%s-proxy", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getProxyRoleName() string {
	return fmt.Sprintf("%s-proxy", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getProxyRoleBindingName() string {
	return fmt.Sprintf("%s-proxy", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getMinecraftServerServiceAccountName() string {
	return fmt.Sprintf("%s-server", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getMinecraftServerRoleName() string {
	return fmt.Sprintf("%s-server", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getMinecraftServerRoleBindingName() string {
	return fmt.Sprintf("%s-server", b.Instance.Name)
}

func (b *MinecraftClusterResourceBuilder) getLabels() map[string]string {
	labels := map[string]string{
		"app.kubernetes.io/name":             b.Instance.Name,
		"app.kubernetes.io/component":        "cluster",
		"minecraftcluster.shulkermc.io/name": b.Instance.Name,
	}

	return labels
}
