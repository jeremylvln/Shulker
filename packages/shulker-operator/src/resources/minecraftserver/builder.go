/*
Copyright (c) Jérémy Levilain
SPDX-License-Identifier: GPL-3.0-or-later
*/

package resources

import (
	"context"
	"fmt"

	shulkermciov1alpha1 "github.com/jeremylvln/shulker/packages/shulker-crds/src/v1alpha1"
	common "github.com/jeremylvln/shulker/packages/shulker-resource-utils/src"
	"k8s.io/apimachinery/pkg/runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

type MinecraftServerResourceBuilder struct {
	Instance *shulkermciov1alpha1.MinecraftServer
	Scheme   *runtime.Scheme
	Client   client.Client
	Ctx      context.Context
}

func (b *MinecraftServerResourceBuilder) ResourceBuilders() ([]common.ResourceBuilder, []common.ResourceBuilder) {
	builders := []common.ResourceBuilder{
		b.MinecraftServerGameServer(),
		b.MinecraftServerConfigMap(),
	}
	dirtyBuilders := []common.ResourceBuilder{}

	if b.Instance.Spec.Configuration.ExistingConfigMapName == "" {
		builders = append(builders, b.MinecraftServerConfigMap())
	}

	return builders, dirtyBuilders
}

func (b *MinecraftServerResourceBuilder) GetGameServerName() string {
	return b.Instance.Name
}

func (b *MinecraftServerResourceBuilder) GetConfigMapName() string {
	if b.Instance.Spec.Configuration.ExistingConfigMapName != "" {
		return b.Instance.Spec.Configuration.ExistingConfigMapName
	}
	return fmt.Sprintf("%s-config", b.Instance.Name)
}

func (b *MinecraftServerResourceBuilder) getServiceAccountName() string {
	return fmt.Sprintf("%s-server", b.Instance.Spec.ClusterRef.Name)
}

func (b *MinecraftServerResourceBuilder) getLabels() map[string]string {
	labels := map[string]string{
		"app.kubernetes.io/name":             b.Instance.Name,
		"app.kubernetes.io/component":        "minecraftserver",
		"minecraftcluster.shulkermc.io/name": b.Instance.Spec.ClusterRef.Name,
	}

	for _, ownerReference := range b.Instance.OwnerReferences {
		if *ownerReference.Controller {
			labels["app.kubernetes.io/name"] = ownerReference.Name
			labels["app.kubernetes.io/instance"] = b.Instance.Name
			labels["minecraftserverfleet.shulkermc.io/name"] = ownerReference.Name
			break
		}
	}

	return labels
}
