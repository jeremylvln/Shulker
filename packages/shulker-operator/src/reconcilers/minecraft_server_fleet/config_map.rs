use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::MinecraftServerFleetReconciler;

pub struct ConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ConfigMapBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = ConfigMap;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        format!("{}-config", minecraft_server_fleet.name_any())
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server_fleet: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            ..ConfigMap::default()
        };

        Ok(config_map)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        config_map: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        config_map.data = Some(
            crate::reconcilers::minecraft_server::config_map::ConfigMapBuilder::get_data_from_spec(
                &minecraft_server_fleet.spec.template.spec.config,
            ),
        );
        Ok(())
    }
}

impl ConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ConfigMapBuilder { client }
    }
}
