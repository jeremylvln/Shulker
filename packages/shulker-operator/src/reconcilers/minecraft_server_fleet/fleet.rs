use std::collections::BTreeMap;

use google_agones_crds::v1::fleet::FleetTemplate;
use k8s_openapi::api::apps::v1::DeploymentStrategy;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use crate::resources::resourceref_resolver::ResourceRefResolver;
use google_agones_crds::v1::fleet::Fleet;
use google_agones_crds::v1::fleet::FleetSpec;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServer;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerSpec;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::config_map::ConfigMapBuilder;
use super::MinecraftServerFleetReconciler;

pub struct FleetBuilder {
    client: Client,
    resourceref_resolver: ResourceRefResolver,
}

#[async_trait::async_trait]
impl ResourceBuilder for FleetBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = Fleet;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        minecraft_server_fleet.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server_fleet: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let fleet = Fleet {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: FleetSpec::default(),
            status: None,
        };

        Ok(fleet)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        fleet: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        let fake_mincraft_server = MinecraftServer {
            metadata: ObjectMeta {
                namespace: minecraft_server_fleet.namespace(),
                name: Some(minecraft_server_fleet.name_any()),
                ..ObjectMeta::default()
            },
            spec: MinecraftServerSpec {
                cluster_ref: minecraft_server_fleet
                    .spec
                    .template
                    .spec
                    .cluster_ref
                    .clone(),
                config: MinecraftServerConfigurationSpec {
                    existing_config_map_name: Some(ConfigMapBuilder::name(minecraft_server_fleet)),
                    ..MinecraftServerConfigurationSpec::default()
                },
                ..minecraft_server_fleet.spec.template.spec.clone()
            },
            status: None,
        };

        let game_server_spec = crate::reconcilers::minecraft_server::gameserver::GameServerBuilder::get_game_server_spec(
            &self.resourceref_resolver,
            &fake_mincraft_server,
        ).await?;
        let replicas = match &minecraft_server_fleet.spec.autoscaling {
            Some(_) => 0,
            None => minecraft_server_fleet.spec.replicas as i32,
        };

        let mut labels = MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet);
        labels.extend(
            game_server_spec
                .template
                .metadata
                .as_ref()
                .map_or_else(BTreeMap::new, |metadata| {
                    metadata.labels.clone().unwrap_or_default()
                }),
        );

        let mut annotations = BTreeMap::<String, String>::new();
        annotations.extend(
            game_server_spec
                .template
                .metadata
                .as_ref()
                .map_or_else(BTreeMap::new, |metadata| {
                    metadata.annotations.clone().unwrap_or_default()
                }),
        );

        if let Some(metadata) = &minecraft_server_fleet.spec.template.metadata {
            if let Some(additional_labels) = metadata.labels.clone() {
                labels.extend(additional_labels);
            }

            if let Some(additional_annotations) = metadata.annotations.clone() {
                annotations.extend(additional_annotations);
            }
        }

        fleet.spec = FleetSpec {
            replicas: Some(replicas),
            strategy: Some(DeploymentStrategy {
                type_: Some("Recreate".to_string()),
                ..DeploymentStrategy::default()
            }),
            scheduling: Some("Packed".to_string()),
            template: FleetTemplate {
                metadata: Some(ObjectMeta {
                    labels: Some(labels),
                    annotations: Some(annotations),
                    ..ObjectMeta::default()
                }),
                spec: game_server_spec,
            },
        };

        Ok(())
    }
}

impl FleetBuilder {
    pub fn new(client: Client) -> Self {
        FleetBuilder {
            client: client.clone(),
            resourceref_resolver: ResourceRefResolver::new(client.clone()),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_server_fleet::fixtures::{create_client_mock, TEST_SERVER_FLEET},
    };

    #[test]
    fn name_contains_fleet_name() {
        // W
        let name = super::FleetBuilder::name(&TEST_SERVER_FLEET);

        // T
        assert_eq!(name, "my-server");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);

        // W
        let fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(fleet);
    }

    #[tokio::test]
    async fn update_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(fleet);
    }

    #[tokio::test]
    async fn update_should_merge_labels() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        let additional_labels = TEST_SERVER_FLEET
            .spec
            .template
            .metadata
            .as_ref()
            .unwrap()
            .labels
            .as_ref()
            .unwrap();
        additional_labels.iter().for_each(|(key, value)| {
            assert_eq!(
                fleet
                    .spec
                    .template
                    .metadata
                    .as_ref()
                    .unwrap()
                    .labels
                    .as_ref()
                    .unwrap()
                    .get(key)
                    .unwrap(),
                value
            );
        });
    }

    #[tokio::test]
    async fn update_should_merge_annotations() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        let additional_annotations = TEST_SERVER_FLEET
            .spec
            .template
            .metadata
            .as_ref()
            .unwrap()
            .annotations
            .as_ref()
            .unwrap();
        additional_annotations.iter().for_each(|(key, value)| {
            assert_eq!(
                fleet
                    .spec
                    .template
                    .metadata
                    .as_ref()
                    .unwrap()
                    .annotations
                    .as_ref()
                    .unwrap()
                    .get(key)
                    .unwrap(),
                value
            );
        });
    }

    #[tokio::test]
    async fn update_should_use_recreate_strategy() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        assert_eq!(
            fleet
                .spec
                .strategy
                .as_ref()
                .unwrap()
                .type_
                .as_ref()
                .unwrap(),
            "Recreate"
        );
    }
}
