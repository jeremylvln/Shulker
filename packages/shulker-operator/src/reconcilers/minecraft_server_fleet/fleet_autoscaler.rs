use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use google_agones_crds::v1::fleet_autoscaler::FleetAutoscaler;
use google_agones_crds::v1::fleet_autoscaler::FleetAutoscalerSpec;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::fleet::FleetBuilder;
use super::MinecraftServerFleetReconciler;

pub struct FleetAutoscalerBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for FleetAutoscalerBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = FleetAutoscaler;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        minecraft_server_fleet.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, minecraft_server_fleet: &Self::OwnerType) -> bool {
        match &minecraft_server_fleet.spec.autoscaling {
            Some(autoscaling) => autoscaling.agones_policy.is_some(),
            None => false,
        }
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let fleet_autoscaler = FleetAutoscaler {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: FleetAutoscalerSpec::default(),
            status: None,
        };

        Ok(fleet_autoscaler)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        fleet_autoscaler: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        fleet_autoscaler.spec = FleetAutoscalerSpec {
            fleet_name: FleetBuilder::name(minecraft_server_fleet),
            policy: minecraft_server_fleet
                .spec
                .autoscaling
                .as_ref()
                .unwrap()
                .agones_policy
                .as_ref()
                .unwrap()
                .clone(),
        };
        Ok(())
    }
}

impl FleetAutoscalerBuilder {
    pub fn new(client: Client) -> Self {
        FleetAutoscalerBuilder { client }
    }
}
