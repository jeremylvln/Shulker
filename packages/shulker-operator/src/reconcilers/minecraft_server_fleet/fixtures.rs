use std::collections::BTreeMap;

use google_agones_crds::v1::fleet_autoscaler::{
    FleetAutoscalerPolicy, FleetAutoscalerPolicyBufferSpec, FleetAutoscalerPolicySpec,
};
use http::{Request, Response};
use k8s_openapi::{api::core::v1::EnvVar, apimachinery::pkg::util::intstr::IntOrString};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    schemas::{FleetAutoscalingSpec, TemplateSpec},
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        minecraft_server::{
            MinecraftServerConfigurationLifecycleStrategy,
            MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
            MinecraftServerPodOverridesSpec, MinecraftServerSpec, MinecraftServerVersion,
            MinecraftServerVersionSpec,
        },
        minecraft_server_fleet::{MinecraftServerFleet, MinecraftServerFleetSpec},
    },
};

lazy_static! {
    pub static ref TEST_SERVER_FLEET: MinecraftServerFleet = MinecraftServerFleet {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-server".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftServerFleetSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            replicas: 3,
            template: TemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(BTreeMap::from([(
                        "test-label/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    annotations: Some(BTreeMap::from([(
                        "test-annotation/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    ..ObjectMeta::default()
                }),
                spec: MinecraftServerSpec {
                    cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
                    tags: vec!["lobby".to_string()],
                    version: MinecraftServerVersionSpec {
                        channel: MinecraftServerVersion::Paper,
                        name: "1.20.1".to_string(),
                        ..MinecraftServerVersionSpec::default()
                    },
                    config: MinecraftServerConfigurationSpec {
                        existing_config_map_name: None,
                        world: Some(ResourceRefSpec {
                            url: Some("https://example.com/my_world.tar.gz".to_string()),
                            ..ResourceRefSpec::default()
                        }),
                        plugins: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_plugin.jar".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        patches: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_patch.tar.gz".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        max_players: 42,
                        disable_nether: false,
                        disable_end: true,
                        server_properties: None,
                        proxy_forwarding_mode:
                            MinecraftServerConfigurationProxyForwardingMode::Velocity,
                        lifecycle_strategy:
                            MinecraftServerConfigurationLifecycleStrategy::AllocateWhenNotEmpty,
                        skip_agent_download: false
                    },
                    pod_overrides: Some(MinecraftServerPodOverridesSpec {
                        image: None,
                        env: Some(vec![EnvVar {
                            name: "EXTRA_ENV".to_string(),
                            value: Some("my_value".to_string()),
                            ..EnvVar::default()
                        }]),
                        resources: None,
                        affinity: None,
                        node_selector: Some(BTreeMap::from([(
                            "beta.kubernetes.io/os".to_string(),
                            "linux".to_string()
                        )])),
                        tolerations: None,
                        service_account_name: None,
                        volume_mounts: None,
                        volumes: None,
                        ports: None,
                        health: None,
                    })
                },
            },
            autoscaling: Some(FleetAutoscalingSpec {
                agones_policy: Some(FleetAutoscalerPolicySpec {
                    type_: FleetAutoscalerPolicy::Buffer,
                    buffer: Some(FleetAutoscalerPolicyBufferSpec {
                        max_replicas: 8,
                        min_replicas: 3,
                        buffer_size: IntOrString::Int(2),
                    }),
                    webhook: None,
                })
            }),
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
