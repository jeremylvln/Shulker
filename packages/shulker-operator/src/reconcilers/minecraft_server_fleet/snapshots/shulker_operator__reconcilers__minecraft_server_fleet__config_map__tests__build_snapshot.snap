---
source: packages/shulker-operator/src/reconcilers/minecraft_server_fleet/config_map.rs
expression: config_map
---
apiVersion: v1
kind: ConfigMap
data:
  bukkit-config.yml: "settings:\n  allow-end: false\nauto-updater:\n  enabled: false\n\n"
  init-fs.sh: "#!/bin/sh\nset -euo pipefail\nset -o xtrace\n\ncp \"${SHULKER_CONFIG_DIR}/server.properties\" \"${SHULKER_SERVER_CONFIG_DIR}/server.properties\"\nif [ \"${SHULKER_VERSION_CHANNEL}\" == \"Paper\" ] || [ \"${SHULKER_VERSION_CHANNEL}\" == \"Folia\" ] || [ \"${SHULKER_VERSION_CHANNEL}\" == \"Minestom\" ]; then\n  cp \"${SHULKER_CONFIG_DIR}/bukkit-config.yml\" \"${SHULKER_SERVER_CONFIG_DIR}/bukkit.yml\"\n  cp \"${SHULKER_CONFIG_DIR}/spigot-config.yml\" \"${SHULKER_SERVER_CONFIG_DIR}/spigot.yml\"\n  mkdir -p \"${SHULKER_SERVER_CONFIG_DIR}/config\"\n  cp \"${SHULKER_CONFIG_DIR}/paper-global-config.yml\" \"${SHULKER_SERVER_CONFIG_DIR}/config/paper-global.yml\"\nfi\n\nif [ ! -z \"${SERVER_WORLD_URL:-}\" ]; then\n  (cd \"${SHULKER_SERVER_CONFIG_DIR}\" && wget \"${SERVER_WORLD_URL}\" -O - | tar -xzv)\nfi\n\nif [ ! -z \"${SHULKER_SERVER_PLUGIN_URLS:-}\" ]; then\n  mkdir -p \"${SHULKER_SERVER_CONFIG_DIR}/plugins\"\n  for plugin_url in ${SHULKER_SERVER_PLUGIN_URLS//;/ }; do\n    (cd \"${SHULKER_SERVER_CONFIG_DIR}/plugins\" && wget \"${plugin_url}\")\n  done\nfi\n\nif [ ! -z \"${SHULKER_SERVER_PATCH_URLS:-}\" ]; then\n  for patch_url in ${SHULKER_SERVER_PATCH_URLS//;/ }; do\n    (cd \"${SHULKER_SERVER_CONFIG_DIR}\" && wget \"${patch_url}\" -O - | tar -xzv)\n  done\nfi\n"
  paper-global-config.yml: "proxies:\n  bungee-cord:\n    online-mode: false\n  velocity:\n    enabled: true\n    online-mode: true\n    secret: ${CFG_VELOCITY_FORWARDING_SECRET}\n\n"
  server.properties: "allow-nether=true\nenforce-secure-profiles=true\nmax-players=42\nonline-mode=false\nprevent-proxy-connections=false\n"
  spigot-config.yml: "settings:\n  bungeecord: false\n  restart-on-crash: false\nadvancements:\n  disable-saving: true\nplayers:\n  disable-saving: true\nstats:\n  disable-saving: true\nsave-user-cache-on-stop-only: true\n\n"
metadata:
  labels:
    app.kubernetes.io/component: minecraft-server
    app.kubernetes.io/instance: config-my-server
    app.kubernetes.io/managed-by: shulker-operator
    app.kubernetes.io/name: config
    app.kubernetes.io/part-of: cluster-my-cluster
    minecraftcluster.shulkermc.io/name: my-cluster
    minecraftserverfleet.shulkermc.io/name: my-server
  name: my-server-config
  namespace: default
